FastSimulation/SimplifiedGeometryPropagator/interface/FastTrackDeDxProducer.h

```diff
@@ -0,0 +1,99 @@
+#ifndef FASTSIM_TRACKRECODEDX_FASTTRACKDEDXPRODUCER_H
+#define FASTSIM_TRACKRECODEDX_FASTTRACKDEDXPRODUCER_H
+
+
+// user include files
+
+#include <memory>
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/TrackerGeometryBuilder/interface/StripGeomDetUnit.h"
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetUnit.h" 
+#include "Geometry/CommonDetUnit/interface/GluedGeomDet.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/TrackReco/interface/DeDxData.h"
+#include "DataFormats/TrackReco/interface/TrackDeDxHits.h"
+#include "DataFormats/TrackReco/interface/DeDxHit.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+
+
+#include "RecoTracker/DeDx/interface/BaseDeDxEstimator.h"
+#include "RecoTracker/DeDx/interface/GenericAverageDeDxEstimator.h"
+#include "RecoTracker/DeDx/interface/TruncatedAverageDeDxEstimator.h"
+#include "RecoTracker/DeDx/interface/MedianDeDxEstimator.h"
+#include "RecoTracker/DeDx/interface/UnbinnedFitDeDxEstimator.h"
+#include "RecoTracker/DeDx/interface/ProductDeDxDiscriminator.h"
+#include "RecoTracker/DeDx/interface/SmirnovDeDxDiscriminator.h"
+#include "RecoTracker/DeDx/interface/ASmirnovDeDxDiscriminator.h"
+#include "RecoTracker/DeDx/interface/BTagLikeDeDxDiscriminator.h"
+
+#include "RecoTracker/DeDx/interface/DeDxTools.h"
+
+#include "TrackingTools/PatternTools/interface/TrajTrackAssociation.h"
+#include "TrackingTools/PatternTools/interface/TrajTrackAssociation.h"
+
+#include "SimDataFormats/TrackingHit/interface/PSimHitContainer.h"
+#include "SimDataFormats/Track/interface/SimTrackContainer.h"
+#include "DataFormats/TrackerRecHit2D/interface/FastTrackerRecHit.h"
+#include "DataFormats/TrackerRecHit2D/interface/FastTrackerRecHitCollection.h"
+
+
+
+//
+// class declaration
+//
+
+class FastTrackDeDxProducer : public edm::stream::EDProducer<> {
+public:
+  explicit FastTrackDeDxProducer(const edm::ParameterSet&);
+  ~FastTrackDeDxProducer() override;
+  static void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+private:
+  void beginRun(edm::Run const& run, const edm::EventSetup&) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  void   makeCalibrationMap(const TrackerGeometry& tkGeom);
+  void   processHit(const FastTrackerRecHit & recHit, float trackMomentum, float& cosine, reco::DeDxHitCollection& dedxHits, int& NClusterSaturating);
+
+  // ----------member data ---------------------------
+BaseDeDxEstimator*               m_estimator;
```

Comment:
```suggestion
  std::unique_ptr<BaseDeDxEstimator> m_estimator;
```
(this comment got lost)

Reply:
this change resulted in a crash since it I guess it is not be consistent with the part of the code that assigns values to m_estimator. Crash:

>> Compiling edm plugin /nfs/dust/cms/user/beinsam/FastSim/4Dec2018/NonDeprecated1/CMSSW_10_4_X_2018-12-12-2300/src/FastSimulation/SimplifiedGeometryPropagator/plugins/FastTrackDeDxProducer.cc 
/nfs/dust/cms/user/beinsam/FastSim/4Dec2018/NonDeprecated1/CMSSW_10_4_X_2018-12-12-2300/src/FastSimulation/SimplifiedGeometryPropagator/plugins/FastTrackDeDxProducer.cc: In constructor 'FastTrackDeDxProducer::FastTrackDeDxProducer(const edm::ParameterSet&)':
/nfs/dust/cms/user/beinsam/FastSim/4Dec2018/NonDeprecated1/CMSSW_10_4_X_2018-12-12-2300/src/FastSimulation/SimplifiedGeometryPropagator/plugins/FastTrackDeDxProducer.cc:63:96: error: no match for 'operator=' (operand types are 'std::unique_ptr<BaseDeDxEstimator>' and 'MedianDeDxEstimator*')
   if     (estimatorName == "median")              m_estimator = new MedianDeDxEstimator(iConfig);

Reply:
done